### ISSUE TITLE
Unexpected TypeLoadException around covariant return and generic base type

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/runtime#47475]
Compile and run the following code:
```
public static class Program_TypeLoadException
{
    public static void Main(string[] args)
    {
        var node = new D<object>();
    }
}

public abstract class C<CT1, CT2>
{
    public C()
    {
        this.This = this;
    }

    protected virtual C<CT1, CT2> This { get; }
}

public class D<DT1> : C<DT1, object>
{
    public D()
    {
        this.This = this;
    }

    protected override D<DT1> This { get; }
}
```

Observed:
```
Unhandled exception. System.TypeLoadException: 
Return type in method 'D`1[DT1].get_This()' on type 'D`1[DT1]' from assembly 'ConsoleAppIssue50649, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' 
is not compatible with base type method 'C`2[DT1,System.__Canon].get_This()'.
   at Program_TypeLoadException.Main(String[] args) 
```

As long as base class of ```D``` refers to ```DT1``` in any position, there is a type load exception. If base class is changed to ```C<object, object>```, there is no exception.

See https://github.com/dotnet/roslyn/issues/50649 for more complicated failing scenarios.

**Comment by AlekseyTs** [comment]
CC @davidwrighton

**Comment by mangod9** [comment]
@janvorli

**Comment by janvorli** [comment]
I cannot repro this issue with the current master branch. The test above passes. From its structure, I believe it was fixed by #47048, which should be in 5.0.3 (porting PR approved #47233). The same is true for the scenario from the roslyn repo.
